[Swift] Use Platform.h feature defines as compilation conditions

emw-apple opened the pull request:
#### dcebb661f44769a670ca7d3bbbd015dd96f63866
<pre>
[Swift] Use Platform.h feature defines as compilation conditions
<a href="https://bugs.webkit.org/show_bug.cgi?id=288691">https://bugs.webkit.org/show_bug.cgi?id=288691</a>
<a href="https://rdar.apple.com/problem/145717488">rdar://problem/145717488</a>

Reviewed by Alexey Proskuryakov.

Reland 291648@main with fixes for production builds. To generate the
response file in install-style builds, reimplement using a build script
phase, and have the generator script loop over all active architectures.

Original commit message:

    Swift can and does import C/C++ headers from WTF, however, it imports
    macro definitions as constant values. Unlike C, swift&apos;s #if expressions
    are parsed separately from module imports, and cannot be influenced by
    imported code. The way to influence compilation conditions is with -D
    arguments, controlled by the SWIFT_ACTIVE_COMPILATION_CONDITIONS build
    setting in Xcode.

    The result of this is that we&apos;ve accumulated a handful of redeclarations
    of out PlatformEnable.h flags in xcconfigs, which have to be carefully
    maintained so that swift compilation matches clang. This change attempts
    to fix this duplication and reestablish Platform.h as the source of
    truth for WebKit&apos;s compile-time feature toggles.

    Add a script to WTF that runs clang&apos;s preprocessor to generate an
    arguments list of enabled features from &lt;wtf/Platform.h&gt;. This is
    similar to what WebCore does in DerivedSources.make, however, it extends
    WebCore&apos;s implementation by being able to solve simple indirection (e.g.
    `#define ENABLE_FOO ENABLE_BAR` will look up the value of ENABLE_BAR).

    Change the default OTHER_SWIFT_FLAGS xcconfig variable to include this
    arguments list as a response file. Swift and Xcode record this as an
    input dependency, so incremental changes to Platform.h and the arguments
    list will invalidate swift code as well.

    Replace some use of SWIFT_ACTIVE_COMPILATION_CONDITIONS with the
    corresponding platform flags. I&apos;ve left WK_ENABLE_SWIFTUI in place,
    because that build setting influences other build settings regarding
    SwiftUI project bringup.

    Swift&apos;s compilation conditions have the same semantics as our ENABLE() /
    HAVE() macros, where an undeclared macro is implicitly undefined, so
    there is no need to come up with a way to make a record of
    deliberately-not-defined features.

* Configurations/CommonBase.xcconfig:
* Configurations/Sanitizers.xcconfig: As a drive-by fix, remove the &quot;=1&quot;
  to silence &quot;Conditional compilation flags do not have values&quot; warning.
* Source/WTF/Scripts/generate-platform-args: Added.
* Source/WTF/WTF.xcodeproj/project.pbxproj: We need to generate one
  response file per compiler target. To do this, add output file
  paths for all current architectures we build for. XCBuild does not
  care whether a declared output actually exists.
* Source/WebKit/Configurations/Base.xcconfig:
* Source/WebKit/UIProcess/API/Cocoa/WebKitSwiftOverlay.swift:
  WebKit_Private doesn&apos;t import successfully on macOS Sonoma, so use a
  compiler guard to only perform the import on newer releases. As a
  drive-by fix, remove a compiler guard that is no longer needed.

Canonical link: <a href="https://commits.webkit.org/291794@main">https://commits.webkit.org/291794@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ba61f4bf3c8c55f53e14eb2e4921f9311aca94d5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94048 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13634 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3377 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/99056 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/44575 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13931 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22065 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/99056 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/44575 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97050 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/10342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/84935 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/99056 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10024 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/2605 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/43891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/86758 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/80273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/2689 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101097 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/92714 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21100 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/15374 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/80769 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21352 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80929 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/80137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19954 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/24689 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/2048 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/14262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21084 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/26262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/115364 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/20771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33145 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/24231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/22512 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Source/WTF/Scripts/generate-platform-args:
```diff
@@ -0,0 +1,91 @@
+import re
+
+# Preprocesses wtf/Platform.h with clang, extracts macro definitions, and saves
+# the platform flags to a Swift-style response file containing -D arguments.
```
Maybe the script name should include "swift" somehow.

emw-apple replied:
There are additional places where we generate listings of platform flags like this (e.g. WebCore below) plus some locations where we compute the search path flags. I'd like to make this script handle those cases as well, in a follow up, so I intended for the script's name to be not-swift-specific.

I'll leave a fixme for those use cases.

aproskuryakov commented on Source/WTF/Scripts/generate-platform-args:
```diff
@@ -0,0 +1,91 @@
+system_header_search_paths = shlex.split(
+    '{SYSTEM_HEADER_SEARCH_PATHS}'.format_map(os.environ))
+preprocessor_definitions = shlex.split(
+    'RELEASE_WITHOUT_OPTIMIZATIONS {GCC_PREPROCESSOR_DEFINITIONS}'.format_map(os.environ))
```
I'm realizing that this is all intricate enough that it would benefit from regression tests. Not saying that you need to add hem before landing, but concerned that this script will be hard to maintain.

Just as an example, I cannot tell why `RELEASE_WITHOUT_OPTIMIZATIONS` is here.

emw-apple replied:
RELEASE_WITHOUT_OPTIMIZATIONS is here because I copied the original logic from DerivedSources.make in WebCore: https://github.com/WebKit/WebKit/blob/01b4cde270584956e84a2caf200480691f4884ce/Source/WebCore/DerivedSources.make#L44-L48

My thought (hope!) is that Swift code will depend on this logic enough that regressions here will cause failures down the line, though specific details like this won't be caught.

aproskuryakov commented on Source/WTF/Scripts/generate-platform-args:
```diff
@@ -0,0 +1,91 @@
+
+    # Resolve macros which are defined as the value of another macro. clang -dM
+    # prints macros in alphabetical order, so we cannot rely on macros being in
+    # topological order. Limit the maximum amount of indirection to prevent a
+    # recursive definition from causing the script to loop forever.
```
Should the script fail with an error when the maximum depth is reached?

emw-apple replied:
I don't think so, because there could be unrelated definitions coming out SDK headers that have no effect on platform defines, and we don't want those to tank the build.

emw-apple committed:
[Swift] Use Platform.h feature defines as compilation conditions
https://bugs.webkit.org/show_bug.cgi?id=288691
rdar://problem/145717488

Reviewed by Alexey Proskuryakov.

Reland 291648@main with fixes for production builds. To generate the
response file in install-style builds, reimplement using a build script
phase, and have the generator script loop over all active architectures.

Original commit message:

    Swift can and does import C/C++ headers from WTF, however, it imports
    macro definitions as constant values. Unlike C, swift's #if expressions
    are parsed separately from module imports, and cannot be influenced by
    imported code. The way to influence compilation conditions is with -D
    arguments, controlled by the SWIFT_ACTIVE_COMPILATION_CONDITIONS build
    setting in Xcode.

    The result of this is that we've accumulated a handful of redeclarations
    of out PlatformEnable.h flags in xcconfigs, which have to be carefully
    maintained so that swift compilation matches clang. This change attempts
    to fix this duplication and reestablish Platform.h as the source of
    truth for WebKit's compile-time feature toggles.

    Add a script to WTF that runs clang's preprocessor to generate an
    arguments list of enabled features from <wtf/Platform.h>. This is
    similar to what WebCore does in DerivedSources.make, however, it extends
    WebCore's implementation by being able to solve simple indirection (e.g.
    `#define ENABLE_FOO ENABLE_BAR` will look up the value of ENABLE_BAR).

    Change the default OTHER_SWIFT_FLAGS xcconfig variable to include this
    arguments list as a response file. Swift and Xcode record this as an
    input dependency, so incremental changes to Platform.h and the arguments
    list will invalidate swift code as well.

    Replace some use of SWIFT_ACTIVE_COMPILATION_CONDITIONS with the
    corresponding platform flags. I've left WK_ENABLE_SWIFTUI in place,
    because that build setting influences other build settings regarding
    SwiftUI project bringup.

    Swift's compilation conditions have the same semantics as our ENABLE() /
    HAVE() macros, where an undeclared macro is implicitly undefined, so
    there is no need to come up with a way to make a record of
    deliberately-not-defined features.

* Configurations/CommonBase.xcconfig:
* Configurations/Sanitizers.xcconfig: As a drive-by fix, remove the "=1"
  to silence "Conditional compilation flags do not have values" warning.
* Source/WTF/Scripts/generate-platform-args: Added.
* Source/WTF/WTF.xcodeproj/project.pbxproj: We need to generate one
  response file per compiler target. To do this, add output file
  paths for all current architectures we build for. XCBuild does not
  care whether a declared output actually exists.
* Source/WebKit/Configurations/Base.xcconfig:
* Source/WebKit/UIProcess/API/Cocoa/WebKitSwiftOverlay.swift:
  WebKit_Private doesn't import successfully on macOS Sonoma, so use a
  compiler guard to only perform the import on newer releases. As a
  drive-by fix, remove a compiler guard that is no longer needed.

Canonical link: https://commits.webkit.org/291794@main
